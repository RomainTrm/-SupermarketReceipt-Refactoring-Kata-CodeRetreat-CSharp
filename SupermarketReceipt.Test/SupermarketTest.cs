using Xunit;
using ApprovalTests;
using ApprovalTests.Reporters;

namespace SupermarketReceipt.Test
{
    [UseReporter(typeof(DiffReporter))]
    public class SupermarketTest
    {
        private static readonly Product Apples = new Product("apples", ProductUnit.Kilo);
        private static readonly Product Toothbrush = new Product("toothbrush", ProductUnit.Each);

        private static Teller CreateTeller()
        {
            SupermarketCatalog catalog = new FakeCatalog();
            catalog.AddProduct(Toothbrush, 0.99);
            catalog.AddProduct(Apples, 1.99);

            return new Teller(catalog);
        }

        [Fact]
        public void WithoutDiscount()
        {
            // ARRANGE
            var cart = new ShoppingCart();
            cart.AddItemQuantity(Apples, 2.5);

            // ACT
            var teller = CreateTeller();
            var receipt = teller.ChecksOutArticlesFrom(cart);

            // ASSERT
            var printReceipt = new ReceiptPrinter().PrintReceipt(receipt);
            Approvals.Verify(printReceipt);
        }

        [Fact]
        public void With10PercentDiscount()
        {
            // ARRANGE
            var cart = new ShoppingCart();
            cart.AddItemQuantity(Apples, 2.5);
            cart.AddItemQuantity(Toothbrush, 3);

            var teller = CreateTeller();
            teller.AddSpecialOffer(SpecialOfferType.TenPercentDiscount, Toothbrush, 10.0);

            // ACT
            var receipt = teller.ChecksOutArticlesFrom(cart);

            // ASSERT
            var printReceipt = new ReceiptPrinter().PrintReceipt(receipt);
            Approvals.Verify(printReceipt);
        }

        [Fact]
        public void WithOfferThreeForTwo()
        {
            // ARRANGE
            var cart = new ShoppingCart();
            cart.AddItemQuantity(Toothbrush, 3);

            var teller = CreateTeller();
            teller.AddSpecialOffer(SpecialOfferType.ThreeForTwo, Toothbrush, 10.0);

            // ACT
            var receipt = teller.ChecksOutArticlesFrom(cart);

            // ASSERT
            var printReceipt = new ReceiptPrinter().PrintReceipt(receipt);
            Approvals.Verify(printReceipt);
        }
        
        [Fact]
        public void WithOfferTwoForAmount()
        {
            // ARRANGE
            var cart = new ShoppingCart();
            cart.AddItemQuantity(Toothbrush, 3);

            var teller = CreateTeller();
            teller.AddSpecialOffer(SpecialOfferType.TwoForAmount, Toothbrush, 10.0);

            // ACT
            var receipt = teller.ChecksOutArticlesFrom(cart);

            // ASSERT
            var printReceipt = new ReceiptPrinter().PrintReceipt(receipt);
            Approvals.Verify(printReceipt);
        }
        
        [Fact]
        public void WithOfferFiveForAmount()
        {
            // ARRANGE
            var cart = new ShoppingCart();
            cart.AddItemQuantity(Toothbrush, 7);

            var teller = CreateTeller();
            teller.AddSpecialOffer(SpecialOfferType.FiveForAmount, Toothbrush, 10.0);

            // ACT
            var receipt = teller.ChecksOutArticlesFrom(cart);

            // ASSERT
            var printReceipt = new ReceiptPrinter().PrintReceipt(receipt);
            Approvals.Verify(printReceipt);
        }
        
        [Fact]
        public void AddMultipleTimesSameProduct()
        {
            // ARRANGE
            var cart = new ShoppingCart();
            cart.AddItemQuantity(Apples, 2.5);
            cart.AddItemQuantity(Apples, 1.5);
            cart.AddItemQuantity(Apples, 3);

            // ACT
            var teller = CreateTeller();
            var receipt = teller.ChecksOutArticlesFrom(cart);

            // ASSERT
            var printReceipt = new ReceiptPrinter().PrintReceipt(receipt);
            Approvals.Verify(printReceipt);
        }
    }
}