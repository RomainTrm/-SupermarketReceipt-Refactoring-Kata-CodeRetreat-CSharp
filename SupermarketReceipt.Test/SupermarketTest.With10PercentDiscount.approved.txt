apples                              4.97
  1.99 * 2.500
toothbrush                          2.97
  0.99 * 3
10% off(toothbrush)                -0.30

Total:                              7.65
